File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int lowestPriceSoFar = prices[0];
        int currentProfit = 0;
        int maxProfit = 0;

        for (int i = 0; i < prices.size() - 1; ++i) {
            currentProfit = prices[i+1] - lowestPriceSoFar;
            lowestPriceSoFar = min(lowestPriceSoFar, prices[i+1]);

            if (currentProfit > maxProfit) {
                maxProfit = currentProfit;
            }
        }

        return maxProfit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        lowestPriceSoFar = prices[0]
        currentProfit = 0
        maxProfit = 0

        for i in range(0, len(prices) - 1):
            currentProfit = prices[i+1] - lowestPriceSoFar
            lowestPriceSoFar = min(lowestPriceSoFar, prices[i+1])

            if (currentProfit > maxProfit):
                maxProfit = currentProfit
        
        return maxProfit
        